### ISSUE TITLE
Issue with Spike Dilutor and Multi-Threading

### ISSUE THREAD

**Issue by wisev** [nest/nest-simulator#2052]
**Describe the bug**
spike_dilutor behavior changes when using multiple threads. 

**To Reproduce**
Here's a small example that exhibits the behavior:
```
import numpy
import nest

# Set Threads https://nest-test.readthedocs.io/en/exhale_test/guides/parallel_computing.html#using-multiple-threads
T=4
nest.SetKernelStatus({"local_num_threads":T})


# Set PRNG seeds https://nest-test.readthedocs.io/en/exhale_test/guides/random_numbers.html#random-numbers-in-parallel-simulations
msd = 123456
N_vp = nest.GetKernelStatus(['total_num_virtual_procs'])[0]
pyrngs = [numpy.random.RandomState(s) for s in range(msd, msd+N_vp)]
nest.SetKernelStatus({'grng_seed' : msd+N_vp})
nest.SetKernelStatus({'rng_seeds' : range(msd+N_vp+1, msd+2*N_vp+1)})

# Define source neurons
neuron_list = []
neuron1 = nest.Create("iaf_psc_delta")
nest.SetStatus(neuron1, {"I_e": 376.0})
neuron_list.append(neuron1)

#Define dilutor with 50% copy
neuron2 = nest.Create("spike_dilutor",params={'p_copy':0.5})

#Create and connect recorders; connect source neurons and dilutor
recorder = nest.Create("spike_recorder")
source_recorder = nest.Create("spike_recorder")

for n in neuron_list:
  nest.Connect(n, neuron2)
  nest.Connect(n, source_recorder)
nest.Connect(neuron2, recorder)

#Run
nest.Simulate(10000)

source_spikes = nest.GetStatus(source_recorder)[0]
spikes = nest.GetStatus(recorder)
print("Original spikes: {}".format(source_spikes))
print("Diluted spikes: {}".format(spikes))
```

- When T = 1, we see 163 source spikes and 77 spikes from the dilutor (expected)
- When T = 2, we see 163 source spikes but 87 spikes from the dilutor
- When T= 4, we see 163 source spikes but 0 spikes from the dilutor
- When T= 8, we see 163 source spikes but 0 spikes from the dilutor

**Expected behavior**
Dilutor should remain consistent, i.e. relay p_copy of the spikes (in this case, 50%).

**Desktop/Environment (please complete the following information):**
 - OS: Ubuntu 20.4.1 LTS
 - Shell: bash
 - Python-Version: 3.8.5
 - NEST-Version: NEST version master@7611bede8, built on Nov 12 2020 17:49:12 for x86_64-pc-linux
 - Installation: Docker

**Additional context**
From the full-scale network, the problem seems to get worse as the number of threads increases.  Issue persists on both Intel Core i9 and AMD Epyc processors.

**Comment by heplesser** [comment]
I can reproduce the bug with the following minimal reproducer. Spikes to and from the dilutor are passed through parrot neurons to avoid any side effects of special handling of spike transmission from/to devices.

```python
import nest

threads = 4
p = 1.0
spikes = [1.0]
n_spk = len(spikes)

nest.ResetKernel()
nest.SetKernelStatus({"local_num_threads": threads})

sg = nest.Create("spike_generator", params={"spike_times": spikes})
p1 = nest.Create("parrot_neuron")
dl = nest.Create("spike_dilutor", params={"p_copy": p})
p2 = nest.Create("parrot_neuron")
sd = nest.Create("spike_recorder")

nest.Connect(sg, p1)
nest.Connect(p1, dl)
nest.Connect(dl, p2)
nest.Connect(p2, sd)

nest.Simulate(10)

print(f"Original spikes: {n_spk}")
print(f"Diluted spikes : {sd.n_events} (expected: {n_spk*p:.0f})")
```

**Comment by heplesser** [comment]
By design, `spike_dilutor` shall send a different diluted spike train to each of its targets. This requires sending spikes through `DSSpikeEvents` and handling them with an `event_hook()`. This again requires that outgoing spikes are sent thread-locally. `spike_dilutor` is therefore replicated on each thread (no proxies).

Currently, `spike_dilutor` is also a `local_receiver`, which means that each replica of the dilutor receives input spikes only from senders on the same thread. Thus, if the node providing input to and the node receiving input from a dilutor "live" on different threads, spikes will never pass through.

What we need to do is to turn `spike_dilutor` into a global receiver and local sender. This is *not* achieved by just changing `local_receiver` to return `false` (seg fault).

**Comment by heplesser** [comment]
@wisev Please excuse the very long delay in following up this issue. After further analysis, I have come to the conclusion that the `spike_dilutor` cannot be fixed for multithreaded simulation. But a solution is available in form of the `bernoulli_synapse`, which does the same thing, but implemented as a synapse instead of as a "pseudo"-neuron.

With the `bernoulli_synapse`, your example can be implemented as follows, updated for NEST 3.3:

```python
import nest

nest.ResetKernel()

nest.local_num_threads = 4
nest.rng_seed = 1234567

neuron = nest.Create("iaf_psc_delta", params={"I_e": 376.0})
parrot = nest.Create("parrot_neuron")

n_rec = nest.Create("spike_recorder")
p_rec = nest.Create("spike_recorder")

nest.Connect(neuron, parrot, syn_spec={'synapse_model': 'bernoulli_synapse', 'p_transmit': 0.5})
nest.Connect(neuron, n_rec)
nest.Connect(parrot, p_rec)

nest.Simulate(10000)

print("Original spikes: {}".format(n_rec.get('n_events')))
print("Diluted spikes : {}".format(p_rec.get('n_events')))
```

This works for any number of threads. The `parrot_neuron` simply passes spikes it receives through. Due to the default delay of 1 ms, spikes emitted by the parrot are 1 ms later than the spikes by the actual neuron.
